From b8c69426ebf8a71c2e895db4b9b641c2705c4725 Mon Sep 17 00:00:00 2001
From: Nox Wilairat <wewilair@microsoft.com>
Date: Thu, 29 Sep 2022 17:56:18 +0000
Subject: [PATCH] Modify CMakeLists.txt to: - build static bsdiff library -
 install bsdiff and bspatch app in the image - install divsufsort divsufsort64
 in the image

---
 CMakeLists.txt | 41 +++++++++++++++++++++++++++++------------
 1 file changed, 29 insertions(+), 12 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index ad3df4b..5797dbc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.10)
+cmake_minimum_required(VERSION 3.20)
 
 project(bsdiff)
 
@@ -27,7 +27,7 @@ endfunction()
 add_libdivsufsort()
 
 # bsdiff
-add_library(bsdiff
+add_library(bsdiff STATIC
     source/bsdiff_private.h
     source/misc.c
     source/file_stream.c
@@ -37,6 +37,7 @@ add_library(bsdiff
     source/bz2_decompress.c
     source/bsdiff.c
     source/bspatch.c)
+
 target_include_directories(bsdiff
     PRIVATE "3rdparty/bzip2"
     PRIVATE "${CMAKE_CURRENT_BINARY_DIR}/3rdparty/libdivsufsort/include"
@@ -50,19 +51,35 @@ endif()
 target_link_libraries(bsdiff PRIVATE bzip2 PRIVATE divsufsort PRIVATE divsufsort64)
 
 if (BUILD_STANDALONES)
-    # bsdiff_diff
-    add_executable(bsdiff_diff source/bsdiff_diff.c)
-    target_include_directories(bsdiff_diff PRIVATE "include")
-    target_link_libraries(bsdiff_diff PRIVATE bsdiff)
+    # bsdiff_app
+    add_executable(bsdiff_app source/bsdiff_diff.c)
+    set_target_properties(bsdiff_app PROPERTIES OUTPUT_NAME "bsdiff")
+    target_include_directories(bsdiff_app PRIVATE "include")
     if (BUILD_SHARED_LIBS)
-        target_compile_definitions(bsdiff_diff PRIVATE "BSDIFF_DLL")
+        target_compile_definitions(bsdiff_app PRIVATE "BSDIFF_DLL")
     endif()
+    target_link_libraries(bsdiff_app PRIVATE bsdiff)
 
-    # bsdiff_patch
-    add_executable(bsdiff_patch source/bsdiff_patch.c)
-    target_include_directories(bsdiff_patch PRIVATE "include")
-    target_link_libraries(bsdiff_patch PRIVATE bsdiff)
+    # bspatch_app
+    add_executable(bspatch_app source/bsdiff_patch.c)
+    set_target_properties(bspatch_app PROPERTIES OUTPUT_NAME "bspatch")
+    target_include_directories(bspatch_app PRIVATE "include")
     if (BUILD_SHARED_LIBS)
-        target_compile_definitions(bsdiff_patch PRIVATE "BSDIFF_DLL")
+        target_compile_definitions(bspatch_app PRIVATE "BSDIFF_DLL")
     endif()
+    target_link_libraries(bspatch_app PRIVATE bsdiff)
+
+    install(TARGETS
+        bsdiff bsdiff_app bspatch_app divsufsort divsufsort64
+        RUNTIME DESTINATION bin
+        LIBRARY DESTINATION lib
+        ARCHIVE DESTINATION lib
+        )
+else()
+    install(TARGETS
+        bsdiff divsufsort divsufsort64
+        RUNTIME DESTINATION bin
+        LIBRARY DESTINATION lib
+        ARCHIVE DESTINATION lib
+        )
 endif()
-- 
2.17.1

